### PR TITLE
Nhan Nguyen - Week 5 Solution

### PR DESCRIPTION
**Pull request by nhannqgh**
- refactor
- apply DI
- apply Clean Architecture

### CODE REVIEW

**Comment by nhannqgh** [comment]
@dathavinh sao pr của mình ko thấy review v bn :((

**Comment by havinhdat** [comment]
@nhaancs , hi, mình sẽ review nốt trong hôm nay nhé, bạn đừng lo 😄

File: core/usecase/get-comments.usecase.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package usecase
+
+import (
+	"github.com/nhaancs/GrabGoTrainingWeek5Assignment/core/repository"
+)
+
+// GetCommentsUsecase usecase
+type GetCommentsUsecase struct {
```
Comment (havinhdat) [inline core/usecase/get-comments.usecase.go]: we should only expose interface, construct
implement struct should be left private

Comment (havinhdat) [inline core/usecase/get-comments.usecase.go]: also name of this struct should be noun, something like Getter could be more descriptive

File: core/usecase/get-comments.usecase.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+// NewGetCommentsUsecase func
+func NewGetCommentsUsecase(repo repository.CommentRepository) *GetCommentsUsecase {
+	return &GetCommentsUsecase {
+		Repo: repo,
+	}
+}
+
+// Execute GetCommentsUsecase
+func (usecase *GetCommentsUsecase) Execute(params... interface{}) (interface{}, error) {
```
Comment (havinhdat) [inline core/usecase/get-comments.usecase.go]: We don't need `params...` here
Nitpick about naming again, this function should be differently, execute express toward a function just do something and return error

File: core/usecase/usecase.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package usecase
+
+// Usecase interface for usecases
+type Usecase interface {
+	Execute(param... interface{}) (interface{}, error)
```
Comment (havinhdat) [inline core/usecase/usecase.go]: We should return something specifically, Golang embrace explicit.

File: data/comment-from-api.data.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+// NewCommentData func
+func NewCommentData(dataformatter dataformatter.DataFormatter) *CommentData {
+	return &CommentData{
+		DataFormatter: dataformatter,
+	}
+}
+
+// GetComments get all comments
+func (data *CommentData) GetComments() ([]entity.Comment, error) {
```
Comment (havinhdat) [inline data/comment-from-api.data.go:23]: Nitpick: this function do 2 things, get data and render it, should have separate into 2 smaller functions

File: main.go [hunk 5]
```diff
@@ -1,121 +1,68 @@
...
-		if err != nil {
-			log.Println("get posts failed with error: ", err)
-			writer.WriteHeader(500)
-			return
-		}
-
-		// Get comments from api
-		comments, err := getComments()
+func handleRequest(rq *requestHandler) {
+	http.HandleFunc(rq.endPoint, func(writer http.ResponseWriter, request *http.Request) {
```
Comment (havinhdat) [inline main.go]: handler function can be write separately from routing

File: main.go [hunk 6]
```diff
@@ -1,121 +1,68 @@
...
+	handleRequest(&requestHandler{
+		endPoint:  "/posts-with-comments",
+		useCase:   getPostsWithCommentsUsecase,
+		formatter: myFormatter,
+	})
 
-	return result
+	log.Println("httpServer starts ListenAndServe at 8080")
+	log.Fatal(http.ListenAndServe(":8080", nil))
 }
```
Comment (havinhdat) [inline main.go:57]: Anyway, thanks for the submission, and sorry for the delay in code-review.
Happy coding!

Comment (nhannqgh) [reply main.go:57]: thanks for your reviews :D

**Pull request closed without merging**